### PR TITLE
adds multithreading

### PR DESCRIPTION
**Pull request by aerickson**
https://bugzilla.mozilla.org/show_bug.cgi?id=1563377

Also:
  - rework how we get statistics about current tests on bitbar

currently running on devicepool0:
78472d4 

previously tried:
c9759531d3c8d5d610493b1e0bba45f33da936c1

### CODE REVIEW

File: mozilla_bitbar_devicepool/test_run_manager.py [hunk 1]
```diff
@@ -88,135 +68,137 @@ def get_bitbar_test_stats(self, project_name, project_config):
...
+    def handle_queue(self, project_name, projects_config):
+        logger.info("thread starting")
+        stats = CACHE['projects'][project_name]['stats']
+
+        while self.state == 'RUNNING':
+            project_config = projects_config[project_name]
+            device_group_name = project_config['device_group_name']
+            additional_parameters = project_config['additional_parameters']
+            worker_type = additional_parameters.get('TC_WORKER_TYPE')
+
```
Comment (bclary) [inline mozilla_bitbar_devicepool/test_run_manager.py]:
```
with lock:
            if stats['OFFLINE'] or stats['DISABLED']:
                ...
                        jobs_to_start))
```

File: mozilla_bitbar_devicepool/test_run_manager.py [hunk 2]
```diff
@@ -88,135 +68,137 @@ def get_bitbar_test_stats(self, project_name, project_config):
...
-            time.sleep(self.wait)
-
-        if self.state == 'TERM':
-            self.abort_tests()
+        # we need the main thread to keep running so it can handle signals
+        # - https://www.g-loaded.eu/2016/11/24/how-to-terminate-running-python-threads-using-signals/
+        while self.state == 'RUNNING':
+            for project_name in projects_config:
+                if project_name == 'defaults':
+                    continue
```
Comment (bclary) [inline mozilla_bitbar_devicepool/test_run_manager.py]:
```
lock = CACHE['projects'][project_name]['lock'] 
with lock:
    self.get_bitbar_test_stats(project_name, projects_config[project_name])
```

File: mozilla_bitbar_devicepool/configuration.py [hunk 3]
```diff
@@ -5,12 +5,14 @@
 from __future__ import absolute_import
 
 import os
```
Comment (bclary) [inline mozilla_bitbar_devicepool/configuration.py]:
```
import threading
```

File: mozilla_bitbar_devicepool/test_run_manager.py [hunk 4]
```diff
@@ -88,135 +68,137 @@ def get_bitbar_test_stats(self, project_name, project_config):
...
+                    ', '.join(stats['OFFLINE_DEVICES'])))
+
+            bitbar_device_group = CACHE['device_groups'][device_group_name]
+            bitbar_device_group_count = bitbar_device_group['deviceCount']
+            taskcluster_provisioner_id = projects_config['defaults']['taskcluster_provisioner_id']
+
+            # create enough tests to service either the pending tasks or twice the number
+            # of the devices in the group (whichever is smaller).
+            pending_tasks = get_taskcluster_pending_tasks(taskcluster_provisioner_id, worker_type)
+            jobs_to_start = min(pending_tasks, 2*bitbar_device_group_count) - stats['WAITING']
```
Comment (bclary) [inline mozilla_bitbar_devicepool/test_run_manager.py]: Let's change this from 2*bitbar_device_group_count to bitbar_device_group_count and just create enough to keep the device group busy but not too busy.

File: mozilla_bitbar_devicepool/configuration.py [hunk 5]
```diff
@@ -138,59 +153,83 @@ def configure_projects():
...
+    project_total = len(projects_config)
+    counter = 0
     for project_name in projects_config:
+        counter += 1
+        log_header = "configure_projects: {} ({}/{})".format(project_name, counter, project_total)
+
         if project_name == 'defaults':
+            logger.info('{}: skipping'.format(log_header))
             continue
+        else:
```
[the remark this answers is not in the document]
Comment (aerickson) [reply mozilla_bitbar_devicepool/configuration.py]: Fixed.

File: mozilla_bitbar_devicepool/configuration.py [hunk 6]
```diff
@@ -203,3 +242,16 @@ def configure_projects():
...
+
+        BITBAR_CACHE['projects'][project_name]['stats'] = {
+            'COUNT': device_group['deviceCount'],
+            'IDLE': 0,
+            'OFFLINE_DEVICES': 0,
+            'OFFLINE': 0,
+            'DISABLED': 0,
+            'RUNNING': 0,
+            'WAITING': 0,
+}
```
[the remark this answers is not in the document]
Comment (aerickson) [reply mozilla_bitbar_devicepool/configuration.py]: Done.

File: mozilla_bitbar_devicepool/test_run_manager.py (hunk 4 above)
Comment (aerickson) [reply mozilla_bitbar_devicepool/test_run_manager.py]: Done.

File: mozilla_bitbar_devicepool/configuration.py (hunk 3 above)
Comment (aerickson) [reply mozilla_bitbar_devicepool/configuration.py]: Why? I can't find any usage of the module in configuration.py.

File: mozilla_bitbar_devicepool/configuration.py [hunk 7]
```diff
@@ -203,3 +242,16 @@ def configure_projects():
             additional_parameters['TASKCLUSTER_ACCESS_TOKEN'] = os.environ[taskcluster_access_token_name]
 
         BITBAR_CACHE['projects'][project_name] = bitbar_project
```
[the remark this answers is not in the document]
Comment (aerickson) [reply mozilla_bitbar_devicepool/configuration.py]: Locking implemented.

**Pull request merged**